File: data_structures/week1_basic_data_structures/tree_height.py
from collections import defaultdict


def get_tree_height(sub_nodes, nodes):
    if not sub_nodes:
        return 1
    max_height = 0
    for node in sub_nodes:
        height = get_tree_height(nodes[node], nodes)
        if height > max_height:
            max_height = height
    return max_height + 1


if __name__ == "__main__":
    n = input()
    elements = [int(_) for _ in input().split(' ')]
    nodes = defaultdict(list)
    for i in range(len(elements)):
        if elements[i] == -1:
            root = i
        else:
            nodes[elements[i]].append(i)

    print(get_tree_height(nodes[root], nodes))

File: data_structures/week1_basic_data_structures/tree_height.go
package main

import (
	"fmt"
)

func getTreeHeight(nodes []int, nodeMap map[int][]int) int {
	height := 1
	for len(nodes) > 0 {
		subNodes := []int{}
		for _, node := range nodes {
			subNodes = append(subNodes, nodeMap[node]...)
		}
		nodes = subNodes
		height++
	}
	return height
}

func main() {
	n, root := 0, 0
	nodeMap := make(map[int][]int)

	fmt.Scanln(&n)
	for i := 0; i < n; i++ {
		node := 0
		fmt.Scanf("%d", &node)
		if node == -1 {
			root = i
		} else {
			nodeMap[node] = append(nodeMap[node], i)
		}
	}
	fmt.Println(getTreeHeight(nodeMap[root], nodeMap))
}
